cpp/util/ManagedArray.h

```diff
@@ -0,0 +1,136 @@
+#ifndef NUMERICAL_ARRAY_H
+#define NUMERICAL_ARRAY_H
+
+#include <memory>
+#include <tbb/tbb.h>
+
+/*! \file ManagedArray.h
+    \brief Defines the standard array class to be used throughout freud.
+*/
+
+namespace freud { namespace util {
+
+//! Class to handle the storage of all arrays of numerical data used in freud.
+/*! The purpose of this class is to handle standard memory management, and to
+ *  provide an abstraction around the implementation-specific choice of
+ *  underlying data structure for arrays of data in freud. These arrays are
+ *  specifically designed for numerical data calculated by a compute class,
+ *  particularly for arrays that must be made accessible through the Python API
+ *  of freud.
+ *
+ *  To support resizing, a ManagedArray instances stores its data as a pointer
+ *  to a pointer, and its size as a pointer. As a result, copy-assignment or
+ *  initialization will result in a new ManagedArray pointing to the same data,
+ *  and any such array can resize or reallocate this data. The pointer to
+ *  pointer infrastructure ensures that such changes properly propagate to all
+ *  ManagedArrays referencing a given memory space. In addition, this
+ *  infrastructure allows the creation of a completely new ManagedArray with a
+ *  new set of pointers that also manages the same data, allowing it to keep
+ *  the original array alive if the original ManagedArray instances become
+ *  decoupled from it.
+ */
+template<typename T> class ManagedArray
+{
+public:
+    //! Default constructor with optional size for thread local arrays
+    /*! \param size Size of the array to allocate.
+     */
+    ManagedArray(unsigned int size=0)
+    {
+        m_size = std::make_shared<unsigned int>(size);
+        m_data = std::shared_ptr<std::shared_ptr<T> >(
+            new std::shared_ptr<T>(new T[size], std::default_delete<T[]>()));
+        reset();
+    }
+
+    //! Destructor (currently empty because data is managed by shared pointer).
+    ~ManagedArray() {}
+
+    //! Update size of the array.
+    /*! \param size New size of the array.
+     */
+    void resize(unsigned int size)
+    {
+        if (size != *m_size)
+        {
+            *m_size = size;
+            reallocate();
```

Comment:
You are correct that the behavior is different. However, when we resize we always wipe the memory because it's fresh computation, that's the standard use-case in freud. We could use a different function name if you prefer, but copying by default on resize is an unnecessary expense.